release_utils/changelog.py

```diff
@@ -0,0 +1,137 @@
+#!/usr/bin/env python
+
+import argparse
+import os
+import json
+import requests
+from subprocess import check_output
+from collections import OrderedDict
+
+
+class ChangeLogIndex(object):
+    """
+    Extracts a list of changelog entries to a text file
+    """
+
+    def cmd(self, c):
+        return check_output(c.split())
+
+    def __init__(self, path):
+        self.path = os.path.abspath(path)
+        conf_file = os.path.join(os.path.dirname(__file__), '../bl.conf')
+        with open(conf_file, 'r') as conf:
+            self.conf = json.load(conf)
+        self.changelogs = OrderedDict()
+        for fname in sorted(
+                self.cmd("find %s -name CHANGELOG" % self.path).split()):
+            with open(fname, 'r') as f:
+                self.changelogs[fname] = f.read().split('\n')
+
+    def match(self, line, tracker):
+        return {"feature": "FEA#", "bug": "BUG#"}.get(tracker) in line
+
+    def get_changelog_entries(self, tracker, version='next'):
+        res = OrderedDict()
+
+        for fname, text in self.changelogs.iteritems():
+            section = None
+            for l in text:
+                if l.strip().startswith("Version "):
+                    section = l.split()[1]
+                    continue
+                if ((version == 'next' and section is None)
+                        or (version == section)) and self.match(l, tracker):
+                    if fname not in res:
+                        res[fname] = []
+                    res[fname].append(l)
+        if not res:
+            print("No entries found for tracker '%s' and version '%s'"
+                % (tracker, version))
+        return res
+
+    def get_redmine_items(self, search_url, item_name):
+        all_items = []
+        resp = requests.get(search_url, auth=(self.conf["api_key"], '')).json()
+        total = resp['total_count']
+        all_items.extend(resp[item_name])
+        offset = 100
+        while len(all_items) < total:
+            resp = requests.get(search_url + '&offset=%s' % offset,
+                    auth=(self.conf["api_key"], '')).json()
+            all_items.extend(resp[item_name])
+            offset += 100
+        return all_items
+
+    def get_features_not_in_changelogs(self):
+        project_id = 1  # Coog
+        status_ids = "3|5"  # resolved, closed
+        tracker_id = 2  # feature
+
+        version_url = self.conf["redmine_url"] + "/projects/coog/versions.json"
+        versions = self.get_redmine_items(version_url, "versions")
+
+        majors = [x for x in versions if x["name"].startswith("Coog-")
+                or x["name"].startswith("Coog ")]
+        major = sorted(majors, key=lambda x: x["id"])[-1]
+        major_number = major["name"].replace(" ", "-").split("-")[1]
+        X, Y = major_number.split(".")
```

Comment:
Tu es sur de ne jamais avoir de "too many values to unpack" ? (genre avec un X.X.X]

Reply:
Si cela arrive c'est qu'on aura changé notre façon de nommer les versions. Il faudra de toute façon changer une bonne partie de cette méthode.